File: packages/shared/src/types/message.ts
export type MessageRole = 'user' | 'assistant' | 'system';

export interface MessageImage {
  path: string;
  filename: string;
}

export interface MessageAttachment {
  path: string;
  filename: string;
  mimeType: string;
  type: 'image' | 'text' | 'pdf' | 'document';
}

export type ChatMediaSource = 'provider' | 'workspace' | 'comfyui' | 'user';

/**
 * Persisted chat media exposed to clients. Storage keys and host filesystem
 * paths intentionally never cross the API boundary.
 */
export interface ChatMedia {
  id: string;
  filename: string;
  mimeType: string;
  byteSize: number;
  altText?: string;
  source: ChatMediaSource;
}

export interface Message {
  id: string;
  sessionId: string;
  /** Thread this message belongs to; null is the implicit legacy main thread. */
  chatId?: string | null;
  role: MessageRole;
  content: string;
  createdAt: string;
  images?: MessageImage[];
  attachments?: MessageAttachment[];
  media?: ChatMedia[];
  /** Stable client id used to reconcile an optimistic/persisted delivery. */
  clientMessageId?: string;
  /** Monotone cursor for replayable events in this session. */
  eventSequence?: number;
}

export interface MessageHistorySnapshot {
  chatId: string | null;
  revision: number;
  highWatermark: number;
  newestMessageId: string | null;
}

export interface MessageJumpTarget {
  sessionId: string;
  chatId: string | null;
  messageId: string;
}

export interface ToolUse {
  name: string;
  input: Record<string, unknown>;
  status: 'pending' | 'running' | 'completed' | 'error';
}

export interface StreamingMessage {
  sessionId: string;
  /** Thread that owns the active provider turn. */
  chatId?: string | null;
  content: string;
  isComplete: boolean;
  toolUse?: ToolUse;
}

File: packages/android/app/src/main/java/com/claudewebui/app/data/model/Message.kt
package com.claudewebui.app.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
enum class MessageRole {
    @SerialName("user") USER,
    @SerialName("assistant") ASSISTANT,
    @SerialName("system") SYSTEM
}

@Serializable
enum class AttachmentType {
    @SerialName("image") IMAGE,
    @SerialName("text") TEXT,
    @SerialName("pdf") PDF,
    @SerialName("document") DOCUMENT
}

@Serializable
data class MessageImage(
    val path: String,
    val filename: String
)

@Serializable
data class MessageAttachment(
    val path: String,
    val filename: String,
    val mimeType: String,
    val type: AttachmentType
)

/**
 * Durable media metadata the REST layer hydrates onto messages. Fetch the bytes
 * from `GET /api/sessions/:sessionId/media/:id`.
 */
@Serializable
data class ChatMedia(
    val id: String,
    val filename: String = "",
    val mimeType: String = "",
    val byteSize: Long = 0,
    val altText: String? = null,
    val source: String? = null
)

@Serializable
data class Message(
    val id: String,
    val sessionId: String,
    /** Thread identity is explicit on new servers; null remains legacy-compatible. */
    val chatId: String? = null,
    val role: MessageRole,
    val content: String,
    val createdAt: String,
    /** What the backend actually sends; see ChatMedia. */
    val media: List<ChatMedia>? = null,
    val images: List<MessageImage>? = null,
    val attachments: List<MessageAttachment>? = null,
    val clientMessageId: String? = null,
    val eventSequence: Long? = null,
)

@Serializable
data class StreamingMessage(
    val sessionId: String,
    val chatId: String? = null,
    val content: String,
    val isComplete: Boolean,
    val toolUse: ToolUseInfo? = null
)

@Serializable
data class ToolUseInfo(
    val name: String,
    val input: kotlinx.serialization.json.JsonElement? = null,
    val status: ToolUseStatus
)

@Serializable
enum class ToolUseStatus {
    @SerialName("pending") PENDING,
    @SerialName("running") RUNNING,
    @SerialName("completed") COMPLETED,
    @SerialName("error") ERROR
}

@Serializable
data class FileAttachmentData(
    val data: String, // base64 encoded
    val mimeType: String,
    val filename: String? = null
)

/**
 * A selected Android document before it is sent. Keeping only its content URI
 * here avoids retaining a second, Base64-expanded copy of large files in
 * Compose state. The bytes are read and encoded immediately before emit.
 */
data class PendingFileAttachment(
    val uri: String,
    val mimeType: String,
    val filename: String,
    val sizeBytes: Long? = null,
)
